aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -35,6 +35,17 @@ template<int n> struct TensorListMetadata
   int block_to_chunk[depth_to_max_blocks[n-1]];
 };
 
+
+template<int n> struct TensorListScalarListMetadata
```

Comment:
After some further back of envelope calculations I think we should be more size-cautious about `TensorListScalarListMetadata` than we are with `TensorListMetadata`.  Let's say TensorListMetadata looks like
```
template<int n, typename T> struct TensorListScalarListMetadata
{
  void* addresses[n][depth_to_max_tensors[n-1]];
  int sizes[depth_to_max_tensors[n-1]];
  T scalar_vals[depth_to_max_tensors[n-1]];
  unsigned char block_to_tensor[depth_to_max_blocks[n-1]];
  int block_to_chunk[depth_to_max_blocks[n-1]];
};
```
If `T` is `double`, and we use `depth_to_max_tensors` as-is, `TensorListScalarListMetadata<1>` requires at least 3800 bytes (+alignment bytes) which is uncomfortably close to the 4KB limit.  I suggest we create a separate heuristic array for `TensorListScalarListMetadata`: 
```
static depth_to_max_tensors_scalarlist = {96, 64, 48, 36, 30};
```
and use `depth_to_max_tensors_scalarlist` in `TensorListScalarListMetadata`.  Changing 110 to 96 gives us a more comfortable high-water mark of 3520 bytes ([I made a gist for quick size testing, try it with nvcc](https://gist.github.com/mcarilli/b704b7af9298d60c70bada798fb64641)) and freedom to tune heuristics for each struct separately in the future.

It also occurs to me we could avoid `TensorListScalarListMetadata` entirely:
Use `TensorListMetadata` with a depth of 2.On the CPU side, pack `tl.addresses[2][:]` with the scalar values cast to `void*`, then in the kernel, retrieve `T y = static_cast<T>(tl.addresses[1][tensor_loc];)`.
I don't think that's as readable though, and `TensorListMetadata<2>` allows only 64 tensors per launch while `TensorListScalarListMetadata` will allow 96. I think writing the separate `TensorListScalarListMetadata` is ok.